File: python/src/storage/adapters/postgresql/transaction.py
"""
PostgreSQL transaction implementation.

This module provides transaction management for PostgreSQL using asyncpg
with support for all isolation levels and savepoints.
"""

from __future__ import annotations

import asyncio
import uuid
from typing import List, Optional

import asyncpg

from storage.adapters.postgresql.result import PostgreSQLResult, PostgreSQLRow
from storage.errors import (
    TransactionError,
    new_transaction_error,
    new_transaction_timeout_error,
)
from storage.interfaces import Transaction
from storage.types import (
    Command,
    ExecuteResult,
    IsolationLevel,
    Operation,
    OperationResult,
    OperationType,
    Query,
    TxOptions,
)


class PostgreSQLTransaction(Transaction):
    """PostgreSQL transaction implementation using asyncpg."""
    
    def __init__(
        self, 
        connection: asyncpg.Connection, 
        pool: asyncpg.Pool,
        options: TxOptions
    ) -> None:
        """Initialize PostgreSQL transaction.
        
        Args:
            connection: Database connection
            pool: Connection pool (for releasing connection)
            options: Transaction options
        """
        self._connection = connection
        self._pool = pool
        self._options = options
        self._transaction: Optional[asyncpg.transaction.Transaction] = None
        self._active = False
        self._id = str(uuid.uuid4())
        self._savepoints: List[str] = []
    
    async def __aenter__(self) -> PostgreSQLTransaction:
        """Async context manager entry."""
        await self._begin()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Async context manager exit."""
        if exc_type is not None:
            await self.rollback()
        else:
            await self.commit()
    
    async def query(self, query: Query) -> PostgreSQLResult:
        """Execute a query within the transaction.
        
        Args:
            query: The query to execute
            
        Returns:
            Query results
            
        Raises:
            QueryError: If query execution fails
            TransactionError: If transaction is not active
        """
        if not self._active:
            raise new_transaction_error("TX_NOT_ACTIVE", "Transaction is not active")
        
        try:
            # Convert ? placeholders to $n format
            sql = self._convert_placeholders(query.sql)
            
            # Execute query with timeout
            if query.timeout:
                result = await asyncio.wait_for(
                    self._connection.fetch(sql, *query.parameters),
                    timeout=query.timeout.total_seconds()
                )
            else:
                result = await self._connection.fetch(sql, *query.parameters)
            
            return PostgreSQLResult(result, self._connection)
            
        except asyncio.TimeoutError as e:
            raise new_transaction_error("QUERY_TIMEOUT", f"Query timeout in transaction: {query.sql[:100]}...") from e
        except Exception as e:
            raise new_transaction_error("QUERY_FAILED", f"Query failed in transaction: {e}") from e
    
    async def query_one(self, query: Query) -> Optional[PostgreSQLRow]:
        """Execute a query and return a single row.
        
        Args:
            query: The query to execute
            
        Returns:
            Single row or None
            
        Raises:
            QueryError: If query execution fails
            TransactionError: If transaction is not active
        """
        result = await self.query(query)
        return await result.fetchone()
    
    async def execute(self, command: Command) -> ExecuteResult:
        """Execute a command within the transaction.
        
        Args:
            command: The command to execute
            
        Returns:
            Execution result
            
        Raises:
            QueryError: If command execution fails
            TransactionError: If transaction is not active
        """
        if not self._active:
            raise new_transaction_error("TX_NOT_ACTIVE", "Transaction is not active")
        
        try:
            # Convert ? placeholders to $n format
            sql = self._convert_placeholders(command.sql)
            
            # Execute command with timeout
            if command.timeout:
                result = await asyncio.wait_for(
                    self._connection.execute(sql, *command.parameters),
                    timeout=command.timeout.total_seconds()
                )
            else:
                result = await self._connection.execute(sql, *command.parameters)
            
            # Parse result to get rows affected
            rows_affected = self._parse_execute_result(result)
            
            return ExecuteResult(rows_affected=rows_affected)
            
        except asyncio.TimeoutError as e:
            raise new_transaction_error("COMMAND_TIMEOUT", f"Command timeout in transaction: {command.sql[:100]}...") from e
        except Exception as e:
            raise new_transaction_error("COMMAND_FAILED", f"Command failed in transaction: {e}") from e
    
    async def batch(self, operations: List[Operation]) -> List[OperationResult]:
        """Execute multiple operations within the transaction.
        
        Args:
            operations: List of operations to execute
            
        Returns:
            List of operation results
            
        Raises:
            QueryError: If any operation fails
            TransactionError: If transaction is not active
        """
        if not self._active:
            raise new_transaction_error("TX_NOT_ACTIVE", "Transaction is not active")
        
        results = []
        
        for i, operation in enumerate(operations):
            try:
                if operation.operation_type == OperationType.QUERY and operation.query:
                    result = await self.query(operation.query)
                    results.append(OperationResult(index=i, result=result))
                elif operation.operation_type == OperationType.COMMAND and operation.command:
                    result = await self.execute(operation.command)
                    results.append(OperationResult(index=i, result=result))
                else:
                    raise ValueError(f"Invalid operation at index {i}")
            except Exception as e:
                results.append(OperationResult(index=i, error=e))
        
        return results
    
    async def commit(self) -> None:
        """Commit the transaction.
        
        Raises:
            TransactionError: If commit fails
        """
        if not self._active:
            raise new_transaction_error("TX_NOT_ACTIVE", "Transaction is not active")
        
        try:
            if self._transaction:
                await self._transaction.commit()
            self._active = False
            
            # Release connection back to pool
            await self._pool.release(self._connection)
            
        except Exception as e:
            raise new_transaction_error("COMMIT_FAILED", f"Failed to commit transaction: {e}") from e
    
    async def rollback(self) -> None:
        """Rollback the transaction.
        
        Raises:
            TransactionError: If rollback fails
        """
        if not self._active:
            return  # Already rolled back or committed
        
        try:
            if self._transaction:
                await self._transaction.rollback()
            self._active = False
            
            # Release connection back to pool
            await self._pool.release(self._connection)
            
        except Exception as e:
            raise new_transaction_error("ROLLBACK_FAILED", f"Failed to rollback transaction: {e}") from e
    
    async def savepoint(self, name: str) -> None:
        """Create a savepoint.
        
        Args:
            name: Savepoint name
            
        Raises:
            TransactionError: If savepoint creation fails
        """
        if not self._active:
            raise new_transaction_error("TX_NOT_ACTIVE", "Transaction is not active")
        
        try:
            await self._connection.execute(f"SAVEPOINT {name}")
            self._savepoints.append(name)
        except Exception as e:
            raise new_transaction_error("SAVEPOINT_FAILED", f"Failed to create savepoint {name}: {e}") from e
    
    async def rollback_to_savepoint(self, name: str) -> None:
        """Rollback to a savepoint.
        
        Args:
            name: Savepoint name
            
        Raises:
            TransactionError: If rollback fails
        """
        if not self._active:
            raise new_transaction_error("TX_NOT_ACTIVE", "Transaction is not active")
        
        if name not in self._savepoints:
            raise new_transaction_error("SAVEPOINT_NOT_FOUND", f"Savepoint {name} not found")
        
        try:
            await self._connection.execute(f"ROLLBACK TO SAVEPOINT {name}")
            
            # Remove savepoints created after this one
            index = self._savepoints.index(name)
            self._savepoints = self._savepoints[:index + 1]
            
        except Exception as e:
            raise new_transaction_error("ROLLBACK_TO_SAVEPOINT_FAILED", f"Failed to rollback to savepoint {name}: {e}") from e
    
    def is_active(self) -> bool:
        """Check if the transaction is active.
        
        Returns:
            True if transaction is active
        """
        return self._active
    
    def get_id(self) -> str:
        """Get the transaction ID.
        
        Returns:
            Transaction ID
        """
        return self._id
    
    async def _begin(self) -> None:
        """Begin the transaction."""
        try:
            # Map isolation level to PostgreSQL syntax
            isolation_map = {
                IsolationLevel.READ_UNCOMMITTED: "READ UNCOMMITTED",
                IsolationLevel.READ_COMMITTED: "READ COMMITTED",
                IsolationLevel.REPEATABLE_READ: "REPEATABLE READ",
                IsolationLevel.SERIALIZABLE: "SERIALIZABLE",
            }
            
            isolation = isolation_map.get(self._options.isolation, "READ COMMITTED")
            
            # Begin transaction with isolation level
            self._transaction = self._connection.transaction(
                isolation=isolation,
                readonly=self._options.read_only,
            )
            
            await self._transaction.start()
            self._active = True
            
            # Set timeout if specified
            if self._options.timeout:
                # PostgreSQL doesn't have per-transaction timeouts,
                # but we could implement this with asyncio.wait_for
                pass
                
        except Exception as e:
            raise new_transaction_error("BEGIN_FAILED", f"Failed to begin transaction: {e}") from e
    
    def _convert_placeholders(self, sql: str) -> str:
        """Convert ? placeholders to $n format for PostgreSQL."""
        param_index = 1
        result = []
        
        for char in sql:
            if char == "?":
                result.append(f"${param_index}")
                param_index += 1
            else:
                result.append(char)
        
        return "".join(result)
    
    def _parse_execute_result(self, result: str) -> int:
        """Parse execute result to get rows affected."""
        # asyncpg returns strings like "INSERT 0 1" or "UPDATE 3"
        if result.startswith("INSERT"):
            parts = result.split()
            return int(parts[2]) if len(parts) >= 3 else 0
        elif result.startswith(("UPDATE", "DELETE")):
            parts = result.split()
            return int(parts[1]) if len(parts) >= 2 else 0
        else:
            return 0

File: internal/infrastructure/adapters/postgres/transaction.go
package postgres

import (
	"context"
	"fmt"
	"sync"

	"github.com/jackc/pgx/v5"
	"github.com/HybridCache.io/storage/pkg/storage"
)

// PostgreSQLTransaction implements the storage.Transaction interface for PostgreSQL
type PostgreSQLTransaction struct {
	tx      pgx.Tx
	adapter *Adapter
	id      string
	active  bool
	mu      sync.RWMutex
}

// Query executes a query within the transaction
func (t *PostgreSQLTransaction) Query(ctx context.Context, query storage.Query) (storage.Result, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	if !t.active {
		return nil, storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	sql, params, err := t.adapter.TranslateQuery(query)
	if err != nil {
		return nil, storage.NewQueryError("QUERY_TRANSLATION_FAILED", err.Error()).WithCause(err)
	}
	
	rows, err := t.tx.Query(ctx, sql, params...)
	if err != nil {
		return nil, storage.NewQueryError("QUERY_EXECUTION_FAILED", err.Error()).WithCause(err)
	}
	
	return &PostgreSQLResult{rows: rows}, nil
}

// QueryOne executes a query and returns a single row within the transaction
func (t *PostgreSQLTransaction) QueryOne(ctx context.Context, query storage.Query) (storage.Row, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	if !t.active {
		return nil, storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	sql, params, err := t.adapter.TranslateQuery(query)
	if err != nil {
		return nil, storage.NewQueryError("QUERY_TRANSLATION_FAILED", err.Error()).WithCause(err)
	}
	
	row := t.tx.QueryRow(ctx, sql, params...)
	return &PostgreSQLRow{row: row}, nil
}

// Execute executes a command within the transaction
func (t *PostgreSQLTransaction) Execute(ctx context.Context, command storage.Command) (storage.ExecuteResult, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	if !t.active {
		return storage.ExecuteResult{}, storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	sql, params, err := t.adapter.TranslateCommand(command)
	if err != nil {
		return storage.ExecuteResult{}, storage.NewQueryError("COMMAND_TRANSLATION_FAILED", err.Error()).WithCause(err)
	}
	
	result, err := t.tx.Exec(ctx, sql, params...)
	if err != nil {
		return storage.ExecuteResult{}, storage.NewQueryError("COMMAND_EXECUTION_FAILED", err.Error()).WithCause(err)
	}
	
	return storage.ExecuteResult{
		RowsAffected: result.RowsAffected(),
		LastInsertID: 0, // PostgreSQL doesn't support LastInsertID
	}, nil
}

// Batch executes multiple operations within the transaction
func (t *PostgreSQLTransaction) Batch(ctx context.Context, operations []storage.Operation) ([]storage.OperationResult, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	if !t.active {
		return nil, storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	results := make([]storage.OperationResult, len(operations))
	
	batch := &pgx.Batch{}
	
	// Add all operations to the batch
	for i, op := range operations {
		switch op.Type {
		case storage.OperationTypeQuery:
			sql, params, err := t.adapter.TranslateQuery(op.Query)
			if err != nil {
				results[i] = storage.OperationResult{
					Index: i,
					Error: storage.NewQueryError("QUERY_TRANSLATION_FAILED", err.Error()).WithCause(err),
				}
				continue
			}
			batch.Queue(sql, params...)
		case storage.OperationTypeCommand:
			sql, params, err := t.adapter.TranslateCommand(op.Command)
			if err != nil {
				results[i] = storage.OperationResult{
					Index: i,
					Error: storage.NewQueryError("COMMAND_TRANSLATION_FAILED", err.Error()).WithCause(err),
				}
				continue
			}
			batch.Queue(sql, params...)
		}
	}
	
	// Execute the batch
	batchResults := t.tx.SendBatch(ctx, batch)
	defer batchResults.Close()
	
	// Process results
	for i, op := range operations {
		if results[i].Error != nil {
			continue // Skip operations that failed translation
		}
		
		switch op.Type {
		case storage.OperationTypeQuery:
			rows, err := batchResults.Query()
			if err != nil {
				results[i] = storage.OperationResult{
					Index: i,
					Error: storage.NewQueryError("BATCH_QUERY_FAILED", err.Error()).WithCause(err),
				}
			} else {
				results[i] = storage.OperationResult{
					Index:  i,
					Result: &PostgreSQLResult{rows: rows},
				}
			}
		case storage.OperationTypeCommand:
			cmdTag, err := batchResults.Exec()
			if err != nil {
				results[i] = storage.OperationResult{
					Index: i,
					Error: storage.NewQueryError("BATCH_COMMAND_FAILED", err.Error()).WithCause(err),
				}
			} else {
				results[i] = storage.OperationResult{
					Index: i,
					Result: storage.ExecuteResult{
						RowsAffected: cmdTag.RowsAffected(),
						LastInsertID: 0,
					},
				}
			}
		}
	}
	
	return results, nil
}

// Commit commits the transaction
func (t *PostgreSQLTransaction) Commit() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	if !t.active {
		return storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	err := t.tx.Commit(context.Background())
	if err != nil {
		return storage.NewTransactionError("COMMIT_FAILED", err.Error()).WithCause(err)
	}
	
	t.active = false
	return nil
}

// Rollback rolls back the transaction
func (t *PostgreSQLTransaction) Rollback() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	if !t.active {
		return storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	err := t.tx.Rollback(context.Background())
	if err != nil {
		return storage.NewTransactionError("ROLLBACK_FAILED", err.Error()).WithCause(err)
	}
	
	t.active = false
	return nil
}

// Savepoint creates a savepoint within the transaction
func (t *PostgreSQLTransaction) Savepoint(name string) error {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	if !t.active {
		return storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	sql := fmt.Sprintf("SAVEPOINT %s", name)
	_, err := t.tx.Exec(context.Background(), sql)
	if err != nil {
		return storage.NewTransactionError("SAVEPOINT_FAILED", err.Error()).WithCause(err)
	}
	
	return nil
}

// RollbackToSavepoint rolls back to a savepoint
func (t *PostgreSQLTransaction) RollbackToSavepoint(name string) error {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	if !t.active {
		return storage.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}
	
	sql := fmt.Sprintf("ROLLBACK TO SAVEPOINT %s", name)
	_, err := t.tx.Exec(context.Background(), sql)
	if err != nil {
		return storage.NewTransactionError("ROLLBACK_TO_SAVEPOINT_FAILED", err.Error()).WithCause(err)
	}
	
	return nil
}

// IsActive returns whether the transaction is active
func (t *PostgreSQLTransaction) IsActive() bool {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.active
}

// ID returns the transaction ID
func (t *PostgreSQLTransaction) ID() string {
	return t.id
}
